### ISSUE TITLE
Make ActorSystem an interface?

### ISSUE THREAD

**Issue by HCanber** [akkadotnet/akka.net#390]
This is a long post, but most of it is glorious ascii art. :)

**TL;DR;** `ActorSystem` is an abstract class. I want to turn it into an interface and create the static class `ActorSystemFactory` that contains `CreateSystem()` (Alternative 2 below).
### Background

Today we have this hierarchy

```
ActorRefFactory
    ↑
ActorSystem
    ↑
ExtendedActorSystem
    ↑
ActorSystemImpl (internal)

```

I would like to create `TestKitActorSystem` that is an `ActorSystem` with some extra members.

```
ActorRefFactory
    ↑
ActorSystem
    ↑
TestKitActorSystem
```

Since some internal tests needs to cast the system to `ExtendedActorSystem` I also need to inherit from `ActorSystemImpl`, so we need a class `TestKitActorSystemImpl` that inherits `ActorSystemImpl` and implements `TestKitActorSystem`.
### Problem

This would have been fine if ActorSystem wasn't an abstract class.

```
                           ActorRefFactory interface
                                  ↑
                             ActorSystem   abstract
                              ↑      ↑
   abstract ExtendedActorSystem      ↑     
                   ↑                 ↑
     class  ActorSystemImpl        TestKitActorSystem interface
                   ↑                 ↑
                 TestKitActorSystemImpl class

```
### Fix, Alternative 1

Turn `ActorSystem` into the interface `IActorSystem`. `ExtendedActorSystem` would also be turned into an interface and  implementations from `ActorSystem` and `ExtendedActorSystem` would be pushed down to `ActorSystemImpl`.
`ActorSystem` would be turned into a static class, since it contains `CreateSystem`.

```
static ActorSystem         ActorRefFactory interface
                                  ↑
                             IActorSystem  interface
                              ↑      ↑
   interface ExtendedActorSystem     ↑     
                   ↑                 ↑
     class  ActorSystemImpl        TestKitActorSystem interface
                   ↑                 ↑
                 TestKitActorSystemImpl class

```

This is a quite a big breaking change. We need to change every place where we now return an `ActorSystem` to return an `IActorSystem`. 
Code that look like this will break:

``` c#
ActorSystem system = GetSystemFromSomewhere();
```
### Fix, Alternative 2

Create a new `ActorSystemFactory` which contains `CreateSystem`. As above turn the others into interfaces, but without renaming `ActorSystem`.

```
static ActorSystemFactory     ActorRefFactory interface
                                     ↑
                                ActorSystem  interface
                                 ↑      ↑
      interface ExtendedActorSystem     ↑     
                      ↑                 ↑
        class  ActorSystemImpl        TestKitActorSystem interface
                      ↑                 ↑
                    TestKitActorSystemImpl class

```

This would mean that all this code will break:

``` c#
var system = ActorSystem.Create(...);
```

This step can be made over two releases. The first introduces `ActorSystemFactory` and marks `ActorSystem.CreateSystem` as obsolete, the second performs the rest. This way, users will have some time to fix issues based on Obsolete-warnings.
### Discussion

I prefer the second alternative. Your thoughts?

Also,
we should think twice about exposing abstract classes in the public api in the future as they restrict us from performing these kind of changes. (Oh give me traits please!)
